activestorage/app/models/active_storage/variant_with_record.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+class ActiveStorage::VariantWithRecord
+  WEB_IMAGE_CONTENT_TYPES = %w[ image/png image/jpeg image/jpg image/gif ]
+
+  attr_reader :blob, :variation
+
+  def initialize(blob, variation)
+    @blob, @variation = blob, ActiveStorage::Variation.wrap(variation)
+  end
+
+  def processed
+    process
+    self
+  end
+
+  def process
+    transform_blob { |image| create_or_find_record(image: image) } unless processed?
+  end
+
+  def processed?
+    record.present?
+  end
+
+  def image
+    record&.image
+  end
+
+  def url(**options)
+    image&.url(**options)
+  end
+
+  alias_method :service_url, :url
+  deprecate service_url: :url
+
+  private
+    def transform_blob
+      blob.open do |input|
+        if blob.content_type.in?(WEB_IMAGE_CONTENT_TYPES)
+          variation.transform(input) do |output|
+            yield io: output, filename: blob.filename, content_type: blob.content_type
+          end
+        else
+          variation.transform(input, format: "png") do |output|
+            yield io: output, filename: "#{blob.filename.base}.png", content_type: "image/png"
+          end
+        end
+      end
+    end
+
+    def create_or_find_record(image:)
+      @record =
+        ActiveRecord::Base.connected_to(role: ActiveRecord::Base.writing_role) do
```

Comment:
May be desirable to extract this method so apps can override to connect to whatever db role. Probably premature.